[ADP-3335] Add "Use case: Centralized Exchanges" to Deposit Wallet (#4513)
input-output-hk/cardano-wallet: lib/customer-deposit-wallet/test/scenario/Test/Scenario/Wallet/Deposit/Exchanges.lhs.md

# Use case: Centralized Exchange

This document describes how a centralized exchange (CEX) can use the Deposit Wallet to

1. Assign an address to a customer ID.
2. Track deposits at this address.
3. Track deposits at all addresses.
4. Create payments to a different wallet.

# Scenarios, Haskell

In this section, we describe the scenarios using Haskell.

```haskell
module Test.Scenario.Wallet.Deposit.Exchanges
    ( scenarioRestore
    , scenarioStart
    , scenarioCreateAddressList
    , scenarioTrackDepositOne
    , scenarioTrackDepositAll
    , scenarioCreatePayment
    ) where

import Prelude

import Cardano.Crypto.Wallet
    ( XPrv
    , XPub
    )
import Cardano.Wallet.Deposit.IO
    ( WalletEnv
    , WalletInstance
    )
import Cardano.Wallet.Deposit.Pure
    ( Customer
    , TxSummary (..)
    , ValueTransfer (..)
    )
import Cardano.Wallet.Deposit.Read
    ( Address
    , Value
    )
import Test.Scenario.Blockchain
    ( ScenarioEnv
    , ada
    , assert
    , payFromFaucet
    , signTx
    , submitTx
    )

import qualified Cardano.Wallet.Deposit.IO as Wallet
import qualified Data.Map as Map
```

> We use a function `depositFundsAt` to model the act where a customer deposits funds at a given address.

Reviewer: ```suggestion
We use a function `depositFundsAt` to model when customer deposits are funded at a given address.
```

Author: Reformulated, but differently.

```haskell
depositFundsAt :: ScenarioEnv -> Address -> Value -> IO ()
depositFundsAt env address value = payFromFaucet env [(address, value)]
```

## 0. Start a Wallet

A `WalletInstance` denotes a mutable wallet that is actively synchronizing to the blockchain, continuously writes its state to a database file, and responds to queries.

> In order to create a fresh wallet, or in order to restore a wallet from its public key all the way from genesis, use the function `withWalletInit`. In addition to the public key, this function expects a number which equals the numerically largest customer ID previously handled with this wallet.

Reviewer: ```suggestion
In order to create a fresh wallet, or in order to restore a wallet from its public key all the way from genesis, use the function `withWalletInit`. In addition to the public key, this function expects a number which equals numerically to the largest customer ID previously handled with this wallet.
```

Author: * "to equal the numerically largest customer ID" — verb
* "is equal to the numerically largest customer ID" — adjective

Both usages of "equal" are acceptable in English. I often use the verb as it plays better with the active voice.

```haskell
scenarioRestore
    :: XPub -> WalletEnv IO -> IO ()
scenarioRestore xpub env = do
    let knownCustomerCount = 127
    Wallet.withWalletInit env xpub knownCustomerCount $ \w -> do
        value <- Wallet.availableBalance w
        assert $ value == ada 0
```

In order to load the wallet state from a database file and resume operation from that state use the function `withWalletLoad`.

```haskell
scenarioStart
    :: WalletEnv IO -> IO ()
scenarioStart env =
    Wallet.withWalletLoad env $ \w -> do
        value <- Wallet.availableBalance w
        assert $ value == ada 0
```

## 1. Assign an address to a customer ID

> The type `Customer` represents a numerical customer ID.

Reviewer: ```suggestion
The type `Customer` is represented by a numeric customer ID.
```

Author: Ah.

* The customer ID represents a `Customer`, not the other way around.

Also 

* The difference between "numeric" and "numerical" appears to be small, but "numeric" seems to be used more often when talking about data types.

Given such a customer ID, the function `createAddress` will create an address and add the association between the customer and this address to the wallet state.

(The mapping from customer ID to address is deterministic and based on the [BIP-32][] address derivation scheme.)

  [bip-32]: https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki

The function `listCustomers` returns the associations between customers and addresses recorded in the current wallet state.

```haskell
scenarioCreateAddressList
    :: WalletInstance -> IO ()
scenarioCreateAddressList w = do
    let customer = 31
    address <- Wallet.createAddress customer w
    customers <- Wallet.listCustomers w
    assert $ (customer, address) `elem` customers
```

## 2. Track deposits at this address

> As soon as an association between customer and address has been added to the wallet state using `createAddress`, the wallet will track deposits made at this address.

Reviewer: ```suggestion
As soon as an association between customer and address has been established to the wallet state using `createAddress`, the wallet will track deposits sent to this address.
```

Author: I prefer "added to the wallet state" over "established in the wallet state",
but I like "made at" less than "sent to".

We use the type `ChainPoint` to specify points on the blockchain — this type uses both a slot number and a block header to uniquely identify a block. We do this in order to allow atomic operations — in the event that the `to` or `from` point are no longer part of the consensus chain, the `getCustomerHistories` functions throws an exception.

The wallet is synchronized to a particular point on the blockchain — use `getWalletTip` to query it.

```haskell
scenarioTrackDepositAll
    :: ScenarioEnv -> WalletInstance -> IO ()
scenarioTrackDepositAll env w = do
    address1 <- Wallet.createAddress customer1 w
    address2 <- Wallet.createAddress customer2 w

    from <- Wallet.getWalletTip w
    depositFundsAt env address1 coin
    depositFundsAt env address2 coin
    depositFundsAt env address1 (coin <> coin)
    to <- Wallet.getWalletTip w

    history <- Wallet.getCustomerHistories (from,to) w
    assert $
        Map.map received history
      ==
        Map.fromList
            [ (customer1, coin <> coin <> coin)
            , (customer2, coin)
            ]
  where
    customer1, customer2 :: Customer
    customer1 = 1
    customer2 = 2
    coin = ada 3

> ```

Reviewer: what about adding the following property at this point? Below `(to1=from2)`
``` 
history1 <-  getCustomerHistories(from1,to1) w
history2 <- getCustomerHistories(from2,to2) w
history3 <- getCustomerHistories(from1,to2) w
history1 union history2 == history3
```
With adequately defined `union`

Author: > what about adding the following property at this point? Below (to1=from2)

I agree that this would be a good unit test and/or needs to be specified in some form.

However, the focus of these scenario tests is documentation / tutorial-style exposition, and I believe that the range argument is documented well enough in text form, so that a reader can infer this property from the text.
